### PR TITLE
Migrate HFDeepSpeedConfig from trfrs to accelerate

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Migrates HFDeepSpeedConfig from transformers repo to accelerate repo as it is generic enough and specific bits remain in transformers.
2. Fix affected code and test; refactoring and improving few things.

### CODE REVIEW

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -215,7 +213,12 @@ class DeepSpeedPlugin:
     This plugin is used to integrate DeepSpeed.
     """
 
-    config_file: str = field(default=None, metadata={"help": "Path to the DeepSpeed config file."})
+    hf_ds_config: Any = field(
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:216]: This is breaking. While the API was marked as experimental and we can do such a change, I'd like to understand why we can't keep the old argument name here.

**Review by sgugger — COMMENTED** [review]
LGTM except the change of argument name. Can we find a way to keep backward compatibility on this?
Also would it affect old accelerate configs of users (in which case we should add a conversion old config to new config to make their lives easier)

File: src/accelerate/utils/dataclasses.py (hunk 1 above)
Comment (pacman100) [reply src/accelerate/utils/dataclasses.py:216]: Hello, this isn't breaking change as this feature was added in PR from couple days ago #405 . It isn't part of the latest release we have and as such won't be a breaking change. The renaming was done to clearly explain that it isn't just limited to file and can accept either file, dict or `HFDeepSpeedConfig` Object.

Comment (sgugger) [reply src/accelerate/utils/dataclasses.py:216]: Oh okay then, I thought it had been there for longer.

File: src/accelerate/state.py [hunk 2]
```diff
@@ -193,3 +194,47 @@ def __repr__(self):
...
+        if isinstance(hf_ds_config, HfDeepSpeedConfig):
+            AcceleratorState._shared_state["deepspeed_plugin"] = DeepSpeedPlugin(
+                hf_ds_config=hf_ds_config, zero3_init_flag=True
+            )
+        else:
+            raise ValueError("`hf_ds_config` must be an instance of `accelerate.utils.HfDeepSpeedConfig`")
+
+    @staticmethod
+    def unset_deepspeed_config():
+        AcceleratorState._shared_state["deepspeed_plugin"] = None
```
Comment (sgugger) [inline src/accelerate/state.py:240]: Let's check if the key exists before setting it a None?

**Review by sgugger — APPROVED** [review]
Nice new utilities! Could you add docstrings on those four new static methods?

File: src/accelerate/state.py [hunk 3]
```diff
@@ -193,3 +194,47 @@ def __repr__(self):
...
+                or ds_config["train_micro_batch_size_per_gpu"] == "auto"
+            ):
+                ds_config["train_micro_batch_size_per_gpu"] = 1
+            if ds_config["train_batch_size"] == "auto":
+                del ds_config["train_batch_size"]
+        return ds_config
+
+    @staticmethod
+    def set_deepspeed_config(hf_ds_config):
+        from .utils import DeepSpeedPlugin, HfDeepSpeedConfig
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/state.py:229]: setting deepspeed_plugin doesn't use any deepspeed functionality; Accelerate and Trainer check DeepSpeed availability at different places. Should we still check deepspeed availability here?

[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/state.py:229]: Oh I thought there was some DeepSpeed import in that file. Okay to leave as is.

File: src/accelerate/state.py (hunk 2 above)
Comment (sgugger) [reply src/accelerate/state.py:240]: As long as you're sure it won't make any deepspeed related stuff break, fine by me.

**Review by sgugger — APPROVED** [review]

**Pull request merged**